beacon_node/store/src/config.rs

```diff
@@ -74,22 +114,88 @@ impl Default for StoreConfig {
 
 impl StoreConfig {
     pub fn as_disk_config(&self) -> OnDiskStoreConfig {
-        OnDiskStoreConfig {
-            slots_per_restore_point: self.slots_per_restore_point,
-        }
+        OnDiskStoreConfig::V22(OnDiskStoreConfigV22::new(self.hierarchy_config.clone()))
     }
 
     pub fn check_compatibility(
         &self,
         on_disk_config: &OnDiskStoreConfig,
+        split: &Split,
+        anchor: &AnchorInfo,
     ) -> Result<(), StoreConfigError> {
-        if self.slots_per_restore_point != on_disk_config.slots_per_restore_point {
-            return Err(StoreConfigError::MismatchedSlotsPerRestorePoint {
-                config: self.slots_per_restore_point,
-                on_disk: on_disk_config.slots_per_restore_point,
-            });
+        // Allow changing the hierarchy exponents if no historic states are stored.
+        let no_historic_states_stored = anchor.no_historic_states_stored(split.slot);
+        let hierarchy_config_changed =
+            if let Ok(on_disk_hierarchy_config) = on_disk_config.hierarchy_config() {
+                *on_disk_hierarchy_config != self.hierarchy_config
+            } else {
+                false
+            };
+
+        if hierarchy_config_changed && !no_historic_states_stored {
+            Err(StoreConfigError::IncompatibleStoreConfig {
+                config: self.as_disk_config(),
+                on_disk: on_disk_config.clone(),
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Check that the configuration is valid.
+    pub fn verify<E: EthSpec>(&self) -> Result<(), StoreConfigError> {
+        self.verify_compression_level()?;
+        self.verify_epochs_per_blob_prune()
+    }
+
+    /// Check that the compression level is valid.
+    fn verify_compression_level(&self) -> Result<(), StoreConfigError> {
+        if zstd::compression_level_range().contains(&self.compression_level) {
+            Ok(())
+        } else {
+            Err(StoreConfigError::InvalidCompressionLevel {
+                level: self.compression_level,
+            })
+        }
+    }
+
+    /// Check that epochs_per_blob_prune is at least 1 epoch to avoid attempting to prune the same
+    /// epochs over and over again.
+    fn verify_epochs_per_blob_prune(&self) -> Result<(), StoreConfigError> {
+        if self.epochs_per_blob_prune > 0 {
+            Ok(())
+        } else {
+            Err(StoreConfigError::ZeroEpochsPerBlobPrune)
+        }
+    }
+
+    /// Estimate the size of `len` bytes after compression at the current compression level.
+    pub fn estimate_compressed_size(&self, len: usize) -> usize {
+        if self.compression_level == 0 {
+            len
+        } else {
+            len / EST_COMPRESSION_FACTOR
```

Comment:
Does the compression factor vary depending on the `compression_level`?

Reply:
No, weirdly the compression level doesn't seem to have any effect :rofl:

Reply:
Added a comment https://github.com/sigp/lighthouse/pull/5978/commits/c4d6ef3544b529bbbc8a919916165e6e976eb1fb